plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java

```diff
@@ -0,0 +1,157 @@
+		QueryAnnotationsEvent.AnnotationFilter filter = new QueryAnnotationsEvent.AnnotationFilter() {
+
+			@Override
+			public boolean accept(Annotation annotation) {
+				return annotation instanceof DiagnosticAnnotation;
+			}
+
+		};
+		QueryAnnotationsEvent.QueryCallback annotationCallback = new QueryAnnotationsEvent.QueryCallback() {
+
+			@Override
+			public void respond(Map<Annotation, Position> annotations) {
+				List<DiagnosticDTO> diagnostics = new ArrayList<>();
+				// iteration with range never returns anything; need to filter ourselves.
+				// https://github.com/eclipse/che/issues/4338
+				annotations.entrySet().stream().filter((e)->e.getValue().overlapsWith(range.getStartOffset(), range.getLength())).forEach((e) -> diagnostics.add(((DiagnosticAnnotation) e.getKey()).getDiagnostic()));
+				CodeActionContextDTO context = CodeActionContextDTOImpl.make().withDiagnostics(diagnostics);
+				CodeActionParamsDTO params = CodeActionParamsDTOImpl.make().withContext(context);
+
+				params.setTextDocument(
+								TextDocumentIdentifierDTOImpl.make().withUri(document.getFile().getLocation().toString()));
+
+				TextPosition start = document.getPositionFromIndex(range.getStartOffset());
+				TextPosition end = document.getPositionFromIndex(range.getStartOffset() + range.getLength());
+				PositionDTO rangeStart = PositionDTOImpl.make().withLine(start.getLine());
+				rangeStart.setCharacter(start.getCharacter());
+				PositionDTO rangeEnd = PositionDTOImpl.make().withLine(end.getLine());
+				rangeEnd.setCharacter(end.getCharacter());
+				RangeDTO rangeParam = RangeDTOImpl.make().withStart(rangeStart);
+				rangeParam.setEnd(rangeEnd);
+				params.setRange(rangeParam);
+
+				Promise<List<CommandDTO>> codeAction = textDocumentService.codeAction(params);
+				List<CompletionProposal> proposals = new ArrayList<>();
+				codeAction.then((commands) -> {
+					for (CommandDTO command : commands) {
+						Action action = actionManager.getAction(command.getCommand());
+						if (action != null) {
+
+							proposals.add(new CompletionProposal() {
+
+								@Override
+								public void getAdditionalProposalInfo(AsyncCallback<Widget> callback) {
+								}
+
+								@Override
+								public String getDisplayString() {
+									return command.getTitle();
+								}
+
+								@Override
+								public Icon getIcon() {
+									return null;
+								}
+
+								@Override
+								public void getCompletion(CompletionCallback callback) {
+									callback.onCompletion(new Completion() {
+
+										@Override
+										public LinearRange getSelection(Document document) {
+											return null;
+										}
+
+										@Override
+										public void apply(Document document) {
+											QuickassistActionEvent evt = new QuickassistActionEvent(new Presentation(), actionManager, perspectiveManager, command.getArguments());
+											action.actionPerformed(evt);
+										}
```

Comment:
I see ten curly brackets almost in a row, can it be somehow refactored to avoid such level of nested code? I can't argue here as I don't know the business logic deep enough but it definitely make it much harder to read suck code. Don't you agree?